README.md
# clojure-weather

Show weather in tray icon.

![screenshot at 2019 01 28 20 13 19 png png image 4480 x 1158 pixels 1](https://user-images.githubusercontent.com/1594701/51853505-8308a280-2339-11e9-8877-0dad7cca7bfd.png)

## Usage
1) Install leiningen (https://github.com/technomancy/leiningen)
2) Change `CITY` constant in the code to your city name.
3) `lein run`


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-01-09
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-01-09
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojure-weather/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure-weather/compare/0.1.0...0.1.1


test/cweather/core_test.clj
(ns clojure-weather.core-test
  (:require [clojure.test :refer :all]
            [clojure-weather.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clojure_weather/core (copy).clj
(ns clojure-weather.core (:gen-class)
  (:use clojure-weather.api)
  (:import [javax.swing JFrame JLabel JTextField JButton]
          [java.awt.event ActionListener]
          [java.awt GridLayout AWTException Color Graphics2D SystemTray TrayIcon Font]
          [java.awt.image BufferedImage])
)

; (defn foo
;   "I don't do a whole lot."
;   [x]
;   (println x "Hello, World!"))

; (defn -main []
;    (println "I'm a little teapot!"))

; (defn -main []
;   (prn "Trying to get location... It will take some time")
;   (let [location "Madrid"]
;     (clojure.pprint/pprint [location (get-temp location)])))

; (defmacro log [& body]
;   `(do
;       (println {"(str (java.util.Date.))" (str (java.util.Date.)) "~(meta &form)" ~(meta &form) "'~body" '~body})
;       (do ~@body)))

; (defmacro kv [arg]
;   `{'~arg ~arg})

(defmacro kv [& args]
  `'~(map #(-> [(str %) %]) args))

(defmacro log [& body]
  `(do
      (println [(kv (str (java.util.Date.))) (kv ~(meta &form)) (kv '~body)])
      (do ~@body)))

; (log (+ 1 2))

; (defmacro kv3 [body]
;   `(do
;       (println ~(str body))))
; (kv3 (cat "1" "2"))

; (kv "1")
; (kv2 "1" "2")



(def ticon nil)

(defn draw-centered-text [g s w h]
  (let [
    fm  (.getFontMetrics g)
    x (int (/ (- w (.stringWidth fm s)) 2))
    y (int (+ (.getAscent fm) (/ (- h (+ (.getAscent fm) (.getDescent fm))) 2)))
    ]
    (prn [w (.stringWidth fm s)])
    (prn [(.getAscent fm) h (.getAscent fm) (.getDescent fm)])
    (prn [g s x y])
    (.drawString g s x y)
  )
)

  ; public void drawCenteredString(String s, int w, int h, Graphics g) {
  ;   FontMetrics fm = g.getFontMetrics();
  ;   int x = (w - fm.stringWidth(s)) / 2;
  ;   int y = (fm.getAscent() + (h - (fm.getAscent() + fm.getDescent())) / 2);
  ;   g.drawString(s, x, y);
  ; }
(let g2 (.createGraphics image))

(defn get-tray-image [tz]
  (let [
    image (new BufferedImage 32 32 (BufferedImage/TYPE_INT_ARGB))
    ]
    ; (.setColor g2 Color/white)
    ; (.fillRect g2 0 0 32 32)
    (.setColor g2 Color/black)
    (.setFont g2 (new Font "TimesRoman" Font/PLAIN, 9))
    ; (.drawString g2 (.toString tz) 0 16)
    (draw-centered-text g2 (.toString tz) 32 32)


    ; e.Graphics.DrawString("My String", FontFamily/GenericSansSerif, Brushes/Black, ClientRectangle, TextFormatFlags.HorizontalCenter |  TextFormatFlags.VerticalCenter | TextFormatFlags.GlyphOverhangPadding);

    ; e.Graphics.DrawString("My String", this.Font, Brushes.Black, ClientRectangle, sf);

    ; g.DrawString(letter, new Font(FontFamily.GenericSansSerif, emSize, FontStyle.Regular),
    ;         new SolidBrush(Color.Black), padx, pady);
    ; FontFamily.GenericSansSerif


    image)
  )

; (defn add-tray-icon [_ tz]
;   (let [
;     systemTray (SystemTray/getSystemTray)
;     trayIcon (new TrayIcon (get-tray-image tz) "Weather Tray")
;     ]
;     (.add systemTray trayIcon))
;   )

(defn add-tray-icon [tz]
  (def ticon (new TrayIcon (get-tray-image tz) "Weather Tray"))
  (.add (SystemTray/getSystemTray) ticon))

(defn update-tray-icon [tz]
  (.setImage ticon (get-tray-image tz)))

(defn p-update-tray-icon [_ tz]
  (update-tray-icon tz))

(defn start-tray-cycle-thread []
  (.start (Thread. (fn [] (
    (loop []
      (log
        (update-tray-icon (get-temp "Moscow"))
        (Thread/sleep 1000)
        (recur))))))))

(defn -setup-ui []
  (let [frame (new JFrame "Weather App")
        location-text (new JTextField "New York")
        main-label (new JLabel "")
        get-button (new JButton "Get Weather")
        temp-label (new JLabel "")
        ]
      (. get-button
          (addActionListener
             (proxy [ActionListener] []
                  (actionPerformed [evt]
                      (.start (Thread. (fn [] (
                        (. get-button (setEnabled false))
                        (. main-label
                             (setText "...Loading..."))
                        (let [temp (get-temp (. location-text (getText)))]
                          (. get-button (setEnabled true))
                          (. main-label
                             (setText temp))
                          (update-tray-icon temp))))))))))
      (doto frame 
                  (.setDefaultCloseOperation (JFrame/EXIT_ON_CLOSE)) ;uncomment this line to quit app on frame close
                  (.setLayout (new GridLayout 2 2 3 3))
                  (.add location-text)
                  (.add main-label)
                  (.add get-button)
                  (.add temp-label)
                  (.setSize 400 200)
                  (.setVisible true))
      (add-tray-icon "N/A")
      (start-tray-cycle-thread)))

(defn -main []
  (-setup-ui)
)


src/clojure_weather/core.clj
(ns clojure-weather.core
  (:gen-class)
  (:import 
          [javax.swing JFrame JLabel JTextField JButton]
          [java.awt.event ActionListener]
          [java.awt GridLayout AWTException Color Graphics2D SystemTray TrayIcon Font PopupMenu MenuItem]
          [java.awt.image BufferedImage])
  (:require
          [clj-http.client :as client]
          [clojure.data.json :as json]))

(def CITY "Berlin")
(def TRAY-SIZE 24)
(def API-KEY "5a043a1bd95bf3ee500eb89de107b41e")

(defmacro to-hash
  "Convert expression into hasmap.
  Example:
  => (to-hash (+ 1 2))
  {(+ 1 2) 3}"
  [arg] `{'~arg ~arg}) 

(defmacro log
  "Log expression to console and execute it.
  Example:
  => (log (+ 1 2))
  [#inst \"2019-01-23T09:02:52.479-00:00\" {:line 1, :column 1} ((+ 1 2))]
  3"
  [& body]
  `(do
      (println [(java.util.Date.) ~(meta &form) '~body])
      (do ~@body)))

(defn get-temp
  ([] (get-temp "New York"))
  ([location]
    (let [url (str "https://api.openweathermap.org/data/2.5/find?q=" location "&units=metric&appid=" API-KEY)]
      (try
        (str
          (-> (client/get url) (get :body) (json/read-str) (get "list") (get 0) (get "main") (get "temp") (str)))
      (catch Exception e
        (prn "Couldn't get weather. Exception caught: " (.getMessage e))
        "N/A")))))

(defn draw-centered-text [g s w h]
  (let [
         fm  (.getFontMetrics g)
         x (int (/ (- w (.stringWidth fm s)) 2))
         y (int (+ (.getAscent fm) (/ (- h (+ (.getAscent fm) (.getDescent fm))) 2)))]
    (.drawString g s x y)))

(defn draw-image [text]
  (let [
         image (new BufferedImage TRAY-SIZE TRAY-SIZE (BufferedImage/TYPE_INT_ARGB))
         g (.createGraphics image)
       ]
    ; (.setColor g Color/white)
    ; (.fillRect g 0 0 32 32)
    (.setColor g Color/black)
    (.setFont g (new Font "DejaVu Sans" Font/PLAIN, 12))
    (draw-centered-text g (.toString text) TRAY-SIZE TRAY-SIZE)
    image))

(defn add-tray-icon [text]
  (let [ticon (new TrayIcon (draw-image text) "Weather Tray")]
    (.add (SystemTray/getSystemTray) ticon)
    ticon))

(defn update-tray-icon [ticon text]
  (.setImage ticon (draw-image text)))

(defn weather-loop [ticon]
  (.start (Thread. (fn [] (
    (loop []
      (log
        (update-tray-icon ticon (get-temp CITY))
        (Thread/sleep (* 1000 60))
        (recur))))))))

(defn menu-item [label callback]
  (let [
         menu (MenuItem. label)
         listener (proxy [ActionListener] [] (actionPerformed [event] (callback)))]
    (.addActionListener menu listener)
    menu))

(defn -main []
  (let [
         frame (new JFrame "Clojure Weather App")
         popup (PopupMenu.)
         ticon (add-tray-icon "N/A")]
    (.add popup (menu-item "Exit" #(System/exit 0)))
    (.setPopupMenu ticon popup)
    (weather-loop ticon)))
